data/modules/MissionUtils/ShipBuilder.lua

```diff
@@ -0,0 +1,855 @@
+function ShipBuilder.MakePlan(template, shipConfig, threat)
+
+	local hullThreat = ShipBuilder.GetHullThreat(shipConfig.id).total
+
+	local randomMod = template.randomModifier or 1.0
+
+	local shipPlan = ShipPlan:clone {
+		threat = hullThreat,
+		freeThreat = threat - hullThreat,
+		maxThreat = threat
+	}
+
+	shipPlan:SetConfig(shipConfig)
+
+	for _, rule in ipairs(template.rules) do
+
+		local canApplyRule = true
+
+		if rule.minThreat then canApplyRule = canApplyRule and threat >= rule.minThreat end
+		if rule.randomChance then canApplyRule = canApplyRule and Engine.rand:Number() < rule.randomChance * randomMod end
+
+		if canApplyRule then
+
+			if rule.slot then
+				ShipBuilder.ApplyEquipmentRule(shipPlan, rule, Engine.rand, hullThreat)
+			else
+				local equip = Equipment.Get(rule.equip)
+				assert(equip)
+
+				local equipThreat = ShipBuilder.ComputeEquipThreatFactor(equip, hullThreat)
+
+				if shipPlan.freeVolume >= equip.volume and shipPlan.freeThreat >= equipThreat then
+					shipPlan:AddEquipToPlan(equip)
+				end
+			end
+
+		end
+
+	end
+
+	shipPlan.label = template.label or Ship.MakeRandomLabel()
+
+	return shipPlan
+
+end
+
+-- Function: ApplyPlan
+--
+-- Apply the previously created plan to a concrete ship object of the
+-- previously-specified hull configuration, installing all equipment instances
+-- created by applying the equipment rules contained in the ship template.
+--
+-- This function ensures that equipment is properly installed in the order it
+-- was selected and items are properly distributed to sub-slots provided by
+-- installed equipment items.
+--
+-- Parameters:
+--
+--  ship     - Ship, the ship to install equipment into.
+--
+--  shipPlan - table, an opaque ship plan returned from a previous call to
+--             ShipBuilder.MakePlan.
+--
+---@param ship Ship
+---@param shipPlan MissionUtils.ShipBuilder.ShipPlan
+function ShipBuilder.ApplyPlan(ship, shipPlan)
+
+	local equipSet = ship:GetComponent('EquipSet')
+
+	-- Apply slot-based equipment first
+	for _, slotId in ipairs(shipPlan.install) do
+		local slot = assert(equipSet:GetSlotHandle(slotId))
+		local equip = assert(shipPlan.filled[slotId])
+		assert(not equip:isProto())
+
+		equipSet:Install(equip, slot)
+	end
+
+	for _, equip in ipairs(shipPlan.equip) do
+		assert(not equip:isProto())
+		equipSet:Install(equip)
+	end
+
+	-- TODO: ammunition / other items inside of instanced equipment
+
+	ship:SetLabel(shipPlan.label)
+
+end
+
+-- =============================================================================
+
+-- Function: MakeShipNear
+--
+-- Spawns a ship near the specified body according to the given ship template
+-- and threat value.
+--
+-- See: Space.SpawnShipNear
+---@param body Body
+---@param template MissionUtils.ShipTemplate
+---@param threat number?
+---@param nearDist number?
+---@param farDist number?
+---@return Ship
+function ShipBuilder.MakeShipNear(body, template, threat, nearDist, farDist)
```

Comment:
This and the following functions are all (almost) identical apart from the function to spawn the actual ship. Any chance to refactor and reduce the copy-pasta?

Reply:
Eh... maybe? The problem is that you need to know the Hull to spawn a concrete ship, and the Plan needs to be generated (and at least loosely validated) before being applied to a ship. At most, you could move the MakePlan -> ApplyPlan step into a separate function, which is... not a significant reduction in the amount of boilerplate.

Reply:
Yeah.. just seemed that these functions are all pretty much the same, but the devil's in the details. I only glanced over this. At least they -are- all pretty much the same, which makes understanding/reviewing easier.

\closing